Source/PluginProcessor.cpp

```diff
@@ -130,21 +124,94 @@ void SoomplerAudioProcessor::processBlock (AudioBuffer<float>& buffer, MidiBuffe
         return;
     }
 
-    auto midiChannelBuffer = filterMidiMessagesForChannel(midiMessages, 1);
+    // TODO maybe this all multi-bus thing is unnesesary?
+//    auto busCount = getBusCount(false);
+//    for (auto busNr = 0; busNr < busCount; ++busNr) {
+//        auto midiChannelBuffer = filterMidiMessagesForChannel(MidiBuffer, busNr + 1);
+//        auto audioBusBuffer = getBusBuffer(buffer, false, busNr);
+
+//        synth.renderNextBlock();
+//    }
+
+    lastMidiEvents = filterMidiMessagesForChannel(midiMessages, 1);
     auto audioBusBuffer = getBusBuffer(buffer, false, 0);
 
-    synth.renderNextBlock(audioBusBuffer, midiChannelBuffer, 0, audioBusBuffer.getNumSamples());
+    synth.renderNextBlock(audioBusBuffer, lastMidiEvents, 0, audioBusBuffer.getNumSamples());
 }
 
 void SoomplerAudioProcessor::processTransport(AudioBuffer<float>& buffer)
 {
-    AudioSourceChannelInfo audioSource{};
+    AudioSourceChannelInfo audioSource;
     audioSource.buffer = &buffer;
     audioSource.startSample = 0;
     audioSource.numSamples = buffer.getNumSamples();
     transportSource.getNextAudioBlock(audioSource);
 }
 
+void SoomplerAudioProcessor::setSampleStartPosition(int64 sample)
+{
+    startSample = sample;
+    transportSource.setNextReadPosition(startSample);
+
+    auto voice = static_cast<soompler::ExtendedVoice*> (synth.getVoice(0));
+    voice->setStartSample(sample);
+}
+
+void SoomplerAudioProcessor::setSampleEndPosition(int64 sample)
+{
+    endSample = sample;
+
+    auto voice = static_cast<soompler::ExtendedVoice*> (synth.getVoice(0));
+    voice->setEndSample(sample);
+}
+
+void SoomplerAudioProcessor::notifyTransportStateChanged(TransportState state)
+{
+    if (transportStateListener == nullptr) {
+        return;
+    }
+
+    transportStateListener->transportStateChanged(state);
+}
+
+std::vector<int> SoomplerAudioProcessor::getActiveNotes()
+{
+    std::vector<int> result;
+
+    // TODO replace when multi-voice synth is realized
+    int noteNumber = synth.getVoice(0)->getCurrentlyPlayingNote();
+
+    if (noteNumber >= 0) {
+        result.push_back(noteNumber);
+    }
+
+    return result;
+}
+
+void SoomplerAudioProcessor::noteOn(int noteNumber)
+{
+    synth.noteOn(0, noteNumber, volume);
+}
+
+void SoomplerAudioProcessor::noteOff(int noteNumber)
+{
+    synth.noteOff(0, noteNumber, volume, true);
+}
+
+MidiBuffer SoomplerAudioProcessor::filterMidiMessagesForChannel(const MidiBuffer &input, int channel)
+{
+    MidiBuffer output;
+    int samplePosition;
+    MidiMessage msg;
+
+    for (MidiBuffer::Iterator it(input); it.getNextEvent(msg, samplePosition);)

```

Comment:
check for using "range based loop" with MidiBuffer

Reply:
There is multiple variables being looped, i don't know right syntax for this